lib.rs

```diff
@@ -93,3 +93,846 @@ pub mod src {
+
+                    if ret.is_again() {
+                        self.pending_data = Some(data);
+                    } else {
+                        dav1d_data_unref(option_nonnull(&mut data));
+                    }
+
+                    return Err(ret);
+                }
+
+                if data.sz > 0 {
+                    self.pending_data = Some(data);
+                    return Err(Error::Again);
+                }
+
+                Ok(())
+            }
+        }
+
+        /// Get the next decoded frame from the decoder.
+        ///
+        /// If this returns `Err([Error::Again])` then further data has to be sent to the decoder
+        /// before further decoded frames become available.
+        ///
+        /// To make most use of frame threading this function should only be called once per submitted
+        /// input frame and not until it returns `Err([Error::Again])`. Calling it in a loop should
+        /// only be done to drain all pending frames at the end.
+        pub fn get_picture(&mut self) -> Result<Picture, Error> {
+            unsafe {
+                let mut pic: Dav1dPicture = mem::zeroed();
+                let ret = dav1d_get_picture(Some(self.dec), option_nonnull(&mut pic));
+
+                if let Err(err) = rav1d_result(ret) {
+                    Err(Error::from(err))
+                } else {
+                    let inner = InnerPicture { pic };
+                    Ok(Picture {
+                        inner: Arc::new(inner),
+                    })
+                }
+            }
+        }
+
+        /// Get the decoder delay.
+        pub fn get_frame_delay(&self) -> u32 {
+            unsafe {
+                dav1d_get_frame_delay(option_nonnull(&self.dec as *const _ as *mut _)).0 as u32
+            }
+        }
+    }
+
+    impl Drop for Decoder {
+        fn drop(&mut self) {
+            unsafe {
+                if let Some(mut pending_data) = self.pending_data.take() {
+                    dav1d_data_unref(option_nonnull(&mut pending_data));
+                }
+                let mut dec = Some(self.dec);
+                dav1d_close(option_nonnull(&mut dec));
+            };
+        }
+    }
+
+    unsafe impl Send for Decoder {}
+    unsafe impl Sync for Decoder {}
+
+    struct InnerPicture {
+        pub pic: Dav1dPicture,
+    }
+
+    /// A decoded frame.
+    #[derive(Clone)]
+    pub struct Picture {
+        inner: Arc<InnerPicture>,
+    }
+
+    /// Pixel layout of a frame.
+    #[derive(Debug, Eq, PartialEq, Copy, Clone)]
+    pub enum PixelLayout {
+        /// Monochrome.
+        I400,
+        /// 4:2:0 planar.
+        I420,
+        /// 4:2:2 planar.
+        I422,
+        /// 4:4:4 planar.
+        I444,
+    }
+
+    /// Frame component.
+    #[derive(Eq, PartialEq, Copy, Clone, Debug)]
+    pub enum PlanarImageComponent {
+        /// Y component.
+        Y,
+        /// U component.
+        U,
+        /// V component.
+        V,
+    }
```

Comment:
These are much better names than we came up with (we don't know much about AV1 or video codecs themselves).  We should try to just use these internally more.